semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+      printf "match%a%a%a%a\n"
+        print_ellipsis ellipsis
+        print_env env
+        print_pat_head pat
+        print_input_head input
+    else if match_calls = max_trace_lines + 1 then
+      printf "[exceeded max trace lines = %i]\n%!" max_trace_lines
+
+module Cache_key = struct
+  type t = ellipsis * env * pattern * input
+
+  let equal (ellipsis1, env1, pat1, input1) (ellipsis2, env2, pat2, input2) =
+    (==) pat1 pat2
+    && (==) input1 input2
+    && ellipsis1 = ellipsis2
+    && Env.equal (=) env1 env2
+
+  let hash_env env =
+    Env.fold
+      (fun k v h ->
+         Hashtbl.hash_param 10 100 k
+         + Hashtbl.hash_param 10 100 v
+         + h)
+      env 0
+
+  (*
+     We define a custom hash function because the default one doesn't work
+     on maps (two equal maps may have different tree structures).
+
+     OCaml's default 'Hashtbl.hash' is the same as 'Hashtbl.hash_param 10 100'.
+     First parameter (10): maximum of number of meaningful nodes used
+                           by the hashing function.
+                           (int-like constants, floats, strings)
+     Second parameter (100): maximum total number of nodes used by the hashing
+                             function.
+     See the documentation for the Hashtbl module for more info.
+  *)
+  let hash (ellipsis, env, pat, input) =
+    Hashtbl.hash_param 10 100 ellipsis
+    + Hashtbl.hash_param 10 100 pat
+    + Hashtbl.hash_param 10 100 input
+    + hash_env env
+end
+
+module Memoize = struct
+  module Tbl = Hashtbl.Make (Cache_key)
+
+  (* only use cache once in 3 times *)
+  let cache_every = 3
+
+  let should_use_cache input =
+    Hashtbl.hash_param 5 10 input mod cache_every = 0
+
+  let get tbl compute ellipsis env pat input =
+    (* only use the cache on some inputs because it's expensive *)
+    if should_use_cache input then
+      let key = (ellipsis, env, pat, input) in
+      match Tbl.find_opt tbl key with
+      | None ->
+          let res = compute ellipsis env pat input in
+          Tbl.add tbl key res;
+          res
+      | Some res -> res
+    else
+      compute ellipsis env pat input
+
+  let create compute =
+    (*
+       Initial table size impact performance.
+    *)
+    let tbl = Tbl.create 8192 in
+    fun ellipsis env pat input ->
+      get tbl compute ellipsis env pat input
+end
+
+(*
+   Main matching function.
+
+   Checks if a pattern matches the entire input sequence.
+   Returns the captured symbols or sequences of symbols for which a
+   name was specified in the pattern.
+*)
+let match_input ?(trace = true) ?(cache = false) root_pat root_input =
+  let stat = { match_calls = 0 } in
+  let get_from_cache = ref (fun _ellipsis _env _pat _input -> assert false) in
+
+  let rec match_ (ellipsis : ellipsis) (env : env) pat input =
+    trace_match_call ~trace stat ellipsis env pat input;
+    !get_from_cache ellipsis env pat input
+
+  and uncached_match ellipsis env pat input =
```

Comment:
would be nice  to have the return type here. What does a matcher return?

Reply:
Apparently either None when no match or a Some with a triplet with X, the environment, and the rest of the input.

Reply:
I clarified this